### PR TITLE
[Model][Qwen] Add Qwen model support.

### PR DESCRIPTION
**Pull request by changqi1**
```bash
$ python ./tools/qwen_convert.py -i /data/qwen-7b-chat-hf/ -o ./qwen-7b-chat-xft --weight_data_type fp32

$ ENABLE_CAT_MLP=0 OMP_NUM_THREADS=48 mpirun -n 1 numactl -N 1 -m 3 ./example --model ../qwen-7b-chat-xft/ --token /data/qwen-7b-chat-hf/tokenizer_config.json --dtype bf16 --loop 1 --no_stream --output_len 16
[INFO] Model path is ../qwen-7b-chat-xft/
[INFO] Token path is /data/qwen-7b-chat-hf/tokenizer_config.json
[INFO] Data type is bf16
[INFO] inputSize is 16
[INFO] outputLen is 16
[INFO] num_beams is 1
[INFO] do_samlpe is false
[INFO] temperature is 1
[INFO] topK is 50
[INFO] topP is 1
[INFO] batch_size is 1
[INFO] loop is 1
[INFO] prefixSharing is OFF
[INFO] Input prompt is : Once upon a time, there existed a little girl who liked to have adventures.
[INFO] Input Token Ids is : 12522 5193 264 882 11 1052 24295 264 2632 3743 879 14915 311 614 30978 13
[INFO] First token time: 145.89 ms
[INFO] Second token time: 71.012 ms

[INFO] Finalzie output is:
==============================================
Once upon a time, there existed a little girl who liked to have adventures. She would often wander off into the woods, exploring and discovering new things. One
```

### CODE REVIEW

File: tools/qwen_convert.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+"""
+Convert huggingface ChatGLM model. Use https://huggingface.co/Qwen
```
Comment (a3213105) [inline tools/qwen_convert.py:2]: maybe using Qwen is better than ChatGLM

Comment (changqi1) [reply tools/qwen_convert.py]: careful 👍

File: examples/cpp/example.cpp [hunk 2]
```diff
@@ -176,6 +177,38 @@ class OptTokenizer : public TokenizerBase {
     const char **vocab_list = vocab_opt;
 };
 
+class QwenTokenizer : public TokenizerBase {
+public:
+    QwenTokenizer(std::string &tokenPath) { vocabSize = 151851; }
+
+    std::vector<int> encode(std::string &input) override {
+        return std::vector<int>({12522, 5193, 264, 882, 11, 1052, 24295, 264, 2632, 3743, 879, 14915, 311, 614, 30978, 13});
```
[the remark this answers is not in the document]
Comment (changqi1) [reply examples/cpp/example.cpp]: Done

File: examples/cpp/example.cpp [hunk 3]
```diff
@@ -176,6 +177,38 @@ class OptTokenizer : public TokenizerBase {
...
+    }
+
+    std::string decode(std::vector<int> &ids) override {
+        if (ids.size() == 1) { return decode(ids[0]); }
+        std::string text("");
```
[the remark this answers is not in the document]
Comment (changqi1) [reply examples/cpp/example.cpp:194]: Done

File: src/models/qwen.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#pragma once
+
+#include "common_decoder.h"
+#include "mlp_llama.h"
+#include "layers_norm.h"
+#include "rotary_embedding.h"
+#include "token_embedding.h"
+
+template <typename WeiT>
+class Qwen : public CommonDecoder<Attention<WeiT, LlamaRotaryEmbedding, RmsNorm>, LlamaMLP<WeiT>, float> {
```
[the remark this answers is not in the document]
Comment (changqi1) [reply src/models/qwen.h:25]: Done

**Review by pujiang2018 — APPROVED** [review]

**Pull request merged**